Add a new debug mode

btw616 opened the pull request:
I have rewritten the code that counts each flowid. It stores each
different (flowid, flowtype, flow_rssbucket) tuple in a red-black
tree. And the counters will be reported when a `SIGINFO` signal is
received (e.g. by pressing `CTRL-T`). This mode (the counting mode)
can be enabled by specifying `-d` option twice, i.e. `-dd`:

```
% ./rss-udp-srv -p 4000 -P 4000 -r 0 -dd
starting: tid=0, rss_bucket=0, cpuset=0
starting: tid=1, rss_bucket=1, cpuset=1
starting: tid=2, rss_bucket=2, cpuset=2
[1] th=0x80181b198
starting: tid=3, rss_bucket=3, cpuset=3
[2] th=0x80181b230
[0] th=0x80181b100
[3] th=0x80181b2c8
starting: tid=4, rss_bucket=4, cpuset=0
starting: tid=5, rss_bucket=5, cpuset=1
[4] th=0x80181b360
starting: tid=6, rss_bucket=6, cpuset=2
starting: tid=7, rss_bucket=7, cpuset=3
[6] th=0x80181b490
[5] th=0x80181b3f8
[7] th=0x80181b528
thr_ev_timer: thr=1, pkts_received=67852, packets_sent=0
thr_ev_timer: thr=1, pkts_received=58535, packets_sent=0
thr_ev_timer: thr=1, pkts_received=67402, packets_sent=0
thr_ev_timer: thr=1, pkts_received=67340, packets_sent=0
thr_ev_timer: thr=1, pkts_received=57496, packets_sent=0
thr_ev_timer: thr=1, pkts_received=67332, packets_sent=0
thr_ev_timer: thr=1, pkts_received=66027, packets_sent=0
thr_ev_timer: thr=1, pkts_received=48016, packets_sent=0
thr_ev_timer: thr=1, pkts_received=70792, packets_sent=0
thr_ev_timer: thr=1, pkts_received=100057, packets_sent=0
thr_ev_timer: thr=1, pkts_received=100779, packets_sent=0
thr_ev_timer: thr=1, pkts_received=125675, packets_sent=0
thr_ev_timer: thr=1, pkts_received=102552, packets_sent=0
thr_ev_timer: thr=1, pkts_received=145, packets_sent=0
load: 0.88  cmd: rss-udp-srv 3803 [kqread] 25.02r 1.79u 5.24s 26% 5448k
count(flowid=0xe4dda387,flowtype=9,flow_rssbucket=7) = 500000
count(flowid=0x3840b491,flowtype=7,flow_rssbucket=1) = 500000
```

The packets are injected by pktgen [1] with following commands:

```
$ sudo ./pktgen -i tap0 -4 -n 500000 -l 100
% sudo ./pktgen -i tap0 -6 -n 500000 -l 100
```

[1] https://github.com/btw616/ip6_direct_test/tree/master/pktgen

Commit:
Add a new debug mode, which can be enabled by specifying "-d" option
twice, i.e. "-dd". It will count each (flowid, flowtype, flow_rssbucket)
tuple, and report the counters when a SIGINFO signal is received.

erikarn commented:
So to properly scale it out, you should keep a tree per thread, and then add them together at the end.
You should also see what the memory bandwidth requirements are for the tree - as long as RB trees don't auto-rebalance every read (ie, you generate memory write traffic whenever you read) it'll be fine.

It may not matter for 2 or 4 CPUs, but it'll matter for 16 CPUs, and that's kind of what we're targeting. :)

btw616 commented:
I got it! I'll optimize my code! ^_^

Commit:
Optimize the code that counts each (flowid, flowtype, flow_rssbucket)
tuple. Now keep a tree per thread, and add them together when needed
to report the counters.

btw616 commented:
I have optimized my code. :-)

Now each thread has a tree, and the counters will be added to a global tree when needed to report the counters.

Commit:
Eliminate the per-tree lock.

Commit:
Disable SIGINFO for udp_srv threads.

erikarn commented:
Thanks!

btw616 commented:
My pleasure! ^_^
